### PR TITLE
ENH: Display git diff when extension update is available

### PR DESCRIPTION
**Pull request by lassoan**
When an extension update is available then a link is added to the extension description to list all the changes between the installed and current version (if the repository is hosted on github and version information is available).

image

When the user clicks the link then the change log is displayed in the default browser:

image

### CODE REVIEW

File: Base/QTGUI/qSlicerExtensionsLocalWidget.cxx [hunk 1]
```diff
@@ -701,15 +710,29 @@ QListWidgetItem* qSlicerExtensionsLocalWidgetPrivate::updateExtensionItem(const
     qWarning() << Q_FUNC_INFO << " failed: missing implementation for serverAPI" << serverAPI;
     }
 
+  QString installedRevision = metadata["installed"].toBool() ? metadata["revision"].toString() : QString();
+  QString onServerRevision = metadataFromServer["revision"].toString();
+
+  // Generate change log URL - only if repository is hosted on github
+  QUrl scmUrl = QUrl(metadataFromServer["scmurl"].toString());
+  QString changeLogUrl;
+  if (scmUrl.host().toLower() == "github.com"
```
Comment (jamesobutler) [inline Base/QTGUI/qSlicerExtensionsLocalWidget.cxx:719]: Should be easy to support the couple of extensions on gitlab.com as well

**Review by jamesobutler — COMMENTED** [review]
Having a quick link to the diff of the commits is a nice added feature 🙏🏻 

On the topic of "Improving user understanding about an available update" I was starting to observe that recently there is a lot more information about an update available to the user which can be overwhelming. There is a lot of info in that blue line.

There is information that a regular user cares about and then there is information that a developer cares about. The info highlighted in red in the image below such as the git hash of the installed version, git hash of the available version and the changelog (reviewing the code diff, since it is not just a release notes) are all information that only a developer will really understand. The alphanumeric git hash is something that is unfamiliar to a non-developer as a version. A non-developer would only understand something like "Version 2.1.0" which is something most extensions do not have so would really just understand the date "2022-07-08" as the version.
image

To address not overshowing info to users that they don't understand, but to still provide info to developers I suggest something maybe like the followingr:
image
- Update button is moved underneath the icon: Currently when there is an update available, as a simple user, I was inclined to press the shiny blue info icon and wasn't immediately drawn to the update button on the far right. It was located between other buttons and was smaller in relative size to the bookmark button. The bookmark button appeared to be more important even when there was an update available. Instead the user would now follow their eye to the shiny extension icons where they would then see the update button.
- Changelog is available, but positioned underneath the update button
- Current installed version number is still displayed in the same position (underneath extension name) with same look and feel as when there is not an available update.
- A tooltip for the Update button would describe the new available version with git hash and date to inspect further. Additionally the download file size would be nice to include in the tooltip as well. This additional info in the update tooltip would reduce the overwhelming amount of additional info to a regular user (non-developer).

File: Base/QTGUI/qSlicerExtensionsLocalWidget.cxx (hunk 1 above)
Comment (lassoan) [reply Base/QTGUI/qSlicerExtensionsLocalWidget.cxx:719]: Good idea. It would be great if you could look it up how it can be done.

**Comment by jamesobutler** [comment]
Without changing the layout of a lot of things at this time, how do we address the issue of too much information about the version update being displayed to the user. The concept of simplifying the display of things for regular users, but providing the more detailed information for more advanced users/developers.

Users may care to know what version is currently installed to provide during debugging. Knowing the version that will be installed however isn't cared by simple users. They will see the new installed version once they install it and can report that if they have issues. Under normal conditions they will just update and not need to check the version of the update. They just care to update when an update is available. Inspecting the new version number or diff log is a more advanced user/developer action.

File: Base/QTGUI/qSlicerExtensionsLocalWidget.cxx (hunk 1 above)
Comment (jamesobutler) [reply Base/QTGUI/qSlicerExtensionsLocalWidget.cxx:719]: The same convention is used for gitlab.

`scmUrl.setPath(QString("%1/compare/%2...%3").arg(scmUrl.path()).arg(installedRevision).arg(onServerRevision));`

For example:
https://gitlab.com/opendose/opendose3d/compare/2d08584afbb71d829d63a00e9f1c7c7898d3b8c5...2a9ebce68627f96da05dda628bada05c3ef827df
which can be used and gets redirected to:
https://gitlab.com/opendose/opendose3d/-/compare/2d08584afbb71d829d63a00e9f1c7c7898d3b8c5...2a9ebce68627f96da05dda628bada05c3ef827df

**Comment by jcfr** [comment]
Rebased for the sake of testing fixes related to apidocs generation - See https://github.com/Slicer/Slicer/issues/4906#issuecomment-1214239522

**Comment by lassoan** [comment]
@jamesobutler I've added gitlab support based on your suggestion and tested everything a bit more. I agree that installed and available versions information could be more subtle, but the version information is already displayed like this in current Slicer (so there is no regression) and implementation of a different layout/behavior would not be trivial (due to the various constraints I described above), so I would not address that in this pull request.

**Pull request merged**